Server/src/services/poi.service.ts

```diff
@@ -135,17 +136,39 @@ export default class POIService {
 	 * @returns percentage of track distance of `poi`, `null` if computation was not possible
 	 */
 	public static async getPOITrackDistancePercentage(poi: POI): Promise<number | null> {
-		// get track distance in kilometers
-		const poiDistKm = await this.getPOITrackDistanceKm(poi)
-		if (poiDistKm == null) {
-			return null
-		}
-
 		// get track length
 		const track = await database.tracks.getById(poi.trackId)
 		if (track == null) {
 			return null
 		}
+
+		// get track distance in kilometers
+		let poiDistKm = await this.getPOITrackDistanceKm(poi)
+		if (poiDistKm == null) {
+			logger.info(`Position of POI with ID ${poi.uid} is not enriched yet.`)
+			// the poi position is not "enriched" yet.
+			// Therefore, obtain and typecast the position
+			const poiPos = GeoJSONUtils.parseGeoJSONFeaturePoint(poi.position);
+			if (poiPos == null){
+				return null
+			}
+			// Then enrich it with the given track,
+			const enrichedPos = await this.enrichPOIPosition(poiPos, track);
+			if (enrichedPos == null){
+				logger.error(`Could not enrich position of POI with ID ${poi.uid}`)
+				return null
+			}
+			// update the poi in the database,
+			poi = await database.pois.update(poi.uid, undefined, undefined, undefined, undefined, enrichedPos) ?? poi
+			// and re-calculate poiDistKm
+			poiDistKm = await this.getPOITrackDistanceKm(poi);
+
+			if (poiDistKm == null){
+				// If we still couldn't, something is weird.
+				logger.error(`Could not get distance of POI with ID ${poi.uid}, even after enriching.`)
+				return null
+			}
+		}
```

Comment:
This whole thing should be done in `getPOITrackDistanceKm` as this function here should only be a wrapper to get the ratio and the other function can be called publicly as well. I will push a commit for that in a few moments.